scenario/mocking.py

```diff
@@ -21,7 +21,7 @@
 )
 
 from ops import JujuVersion, pebble
-from ops.model import ModelError, RelationNotFoundError
+from ops.model import CloudSpec, ModelError, RelationNotFoundError
```

Comment:
We had a discussion recently @tonyandrewmeyer and I about whether we should be using directly objects from `ops` in `State`, and IIRC the answer was no.
Main reason is State needs to be dataclass-serializable and in this case that's luckily the case as CloudSpec is a frozen dataclass like the rest of State. However that is an ops implementation detail so far as scenario is concerned, not sure we should be relying on that. Instead, we'd be safer to reimplement CloudSpec in `scenario.state`. That gives us the freedom to add helper methods and evolve the API independently too.

Big IMHO on this one though, perhaps you folks have already decided you want to make the two more interdependent.

If we agree on the approach I suggest, I'll add a test to assert that the whole State tree is a frozen dataclass or a primitive type.

Reply:
> We had a discussion recently @tonyandrewmeyer and I about whether we should be using directly objects from `ops` in `State`, and IIRC the answer was no.

Yeah, I (and @benhoyt) were originally in favour of it, but when I tried to do this with notices I found it it's actually much messier than I expected and it doesn't offer a lot of benefits, so we (now) both agree that it's not worth doing.

I think we do need to think a bit more in ops sometimes when adding dataclasses about the order of arguments and when things are better as kwarg only. But that's more of a future thing that could make the ops/Scenario curve nicer.

> Main reason is State needs to be dataclass-serializable and in this case that's luckily the case as CloudSpec is a frozen dataclass like the rest of State. However that is an ops implementation detail so far as scenario is concerned, not sure we should be relying on that. Instead, we'd be safer to reimplement CloudSpec in `scenario.state`. That gives us the freedom to add helper methods and evolve the API independently too.

Yeah, I feel like we can get past the serialisation issues, but those two latter benefits are significant. It also makes it easier to have sensible defaults that make sense in a testing context but not in a production one (probably not so relevant with CloudSpec).

I think using enums from ops is fine, and I think the way that some of the pebble objects are used in `Container` (services, layers) is fine too - but as top-level Scenario objects that you actually add to the State, those should all be Scenario-specific classes.